### PR TITLE
[WIP] Rollup build

### PR DESCRIPTION
**Pull request by gaearon**
Just opening PR so we can keep track of progress.
I'll close this one before merging anything so that @trueadm gets the squashed commit 😉 

@trueadm Feel free to reassign these as you see fit.

- [x] Set up a basic build process @trueadm 
- [x] Add forwarding modules @trueadm 
- [x] Fix ReactPerf circular dependency with ReactGKJSModule @gaearon
- [x] Fix CurrentOwner exception
- [x] Fix "addComponentAsRefTo(...): Only a ReactOwner can have refs." @trueadm 
- [x] Figure out a fix for Rollup flat bundle breaking www @trueadm https://github.com/rollup/rollup-plugin-commonjs/issues/176
- [x] Fix dynamic injection breaking because forwarding modules don't know which bundle to refer to @gaearon
- [x] Clean up before build and add a shortcut for building one bundle type @trueadm 
- [x] Fix all missing static resources @gaearon
- [x] Figure out how to inject feature flags into the bundle correctly @gaearon 
- [x] Figure out how to inject error utils and Fiber error dialog @gaearon
- [x] Add a ReactART build @gaearon
- [x] Add a ReactDOMServer build @trueadm 
- [x] Rebase www now that Sebastian landed the sync @gaearon
- [x] CommonJS bundles shouldn't contain `DEV` code (for prod) or `process.env.NODE_ENV` checks since those are slow. We should have separate CommonJS builds for development and production, and switch based on `NODE_ENV` between them. @trueadm 
- [x] Fork addons internally @gaearon
- [x] Make build handle NPM packages, facebook-www and UMD dist @trueadm
- [x] Remove `-core` indirection internally @gaearon
- [x] Figure out a plan for responder/tap plugins, `reactComponentExpect` etc where we'd still like to be the source of truth
- [x] Fix KB profiler not to rely on internals @trueadm
- [x] Figure out how to remove ReactPerf and ReactTestUtils (especially shallow renderer) from www bundles in prod mode
- [x] Check for duplicated output and do some initial profiling @trueadm 
- [x] Ensure we have license headers both for our code and `object-assign` in all bundles @trueadm 
- [x] Delete `examples`, move `examples/fiber` to `fixtures`, and check docs for links to examples
- [x] Figure out the plan for ReactDOMServer bundle
- [x] Fix DOM fixtures and fiber debugger
- [x] Fix ART UMD build
- [x] Add ReactDOMServerStack CommonJS entry point to `react-dom/server`
- [x] Do initial work for RN flat bundle (the rest is follow up) 
- [x] Figure out the plan for TapEventPlugin and friends
- [ ] Remove dependency on Node 7 (we only have Node 4 on devservers)
- [ ] Update www sync script

### CODE REVIEW

**Comment by gaearon** [comment]
This is the fix for ReactPerf circular dep: https://github.com/facebook/react/pull/9147/commits/723b402c07116a70ce8ff1e43a1f4d92052e8f43.

Since https://github.com/rollup/rollup-plugin-commonjs/issues/176 was blocking me, I fixed up the bundles by hand: https://gist.github.com/gaearon/ff710198c82f60470c8c4092b9204065.

Right now I'm stuck on an error related to `ReactCurrentOwner` being broken. You can see my progress in D4686612.

**Comment by Andarist** [comment]
@gaearon 
What are benefits of flat bundle? cant find much about this over the internet. 

Also the description of what have been done here (or will be done) would be great for informative purposes. Just like a high level description, not per file or anything of course.

**Comment by gaearon** [comment]
We'll write something up when it works 😉

File: scripts/rollup/mangle.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+// these are added as they are handled by rollup in some cases
+const propertyMangleWhitelist = [
+  // React
```
Comment (gaearon) [inline scripts/rollup/mangle.js]: Can you explain more about this change and why it is safe? I would prefer that we don't alter the build significantly in this PR, and pursue more aggressive strategies separately.

File: scripts/rollup/build.js [hunk 2]
```diff
@@ -58,6 +60,7 @@ function getAliases(paths, bundleType, isRenderer) {
 
 // the facebook-www directory
 const facebookWWW = 'facebook-www';
+const enablePropertyMangling = false;
```
Comment (gaearon) [inline scripts/rollup/build.js]: Oh I just realized it's disabled by default. 😄

Comment (trueadm) [inline scripts/rollup/build.js]: This stuff is WIP, it's disabled by default too :)

File: scripts/rollup/mangle.js (hunk 1 above)
Comment (trueadm) [inline scripts/rollup/mangle.js]: It's not going to happen yet, I just added support to the tooling to be used when we get around to it. We need to change some things on master for this stuff to work anyway.

**Comment by flarnie** [comment]
You could note in the final PR that this closes https://github.com/facebook/react/issues/4927

**Pull request closed without merging**